### ISSUE TITLE
Align no_unused_private_class_members and use_readonly_class_properties behaviour

### ISSUE THREAD

**Issue by vladimir-ivanov** [biomejs/biome#7499]
### Description

currently no_unused_private_class_members is based on an eslint rule, whereas use_readonly_class_properties is used on tsconfig/eslint rule.
The definitions of reads and writes of a class members and accessors (get, set) within a class are different in some aspects.

Would be nice to have full consistency and code reuse on those.

Another aspect to look at / maybe as a separate task - the potential conflicts if both rules are enabled and auto fixing.

**Comment by vladimir-ivanov** [comment]
my suggestion on this one is to use typescript/eslint rule as the preferred one.

**Comment by vladimir-ivanov** [comment]
with regards to precedence - no unused privatee class members should happen first, and only if they are used we check for readonly.

**Comment by vladimir-ivanov** [comment]
with regards to no unused private class members - simplest rule to start with would be no writes and no reads. 
e.g. semanticClass ctx.references.reads and ctx.references.writes.
From there we can define exceptions if we feel they are not good enough AND / Or modify what is classed as read and write in the actual class itself.

Will try to come up with examples of all conflicting rules shortly in a pr

**Comment by vladimir-ivanov** [comment]
currently failing examples after applying the reads and writes rule ([code]: https://github.com/biomejs/biome/pull/7515/files#diff-bd8ad9c06a551f8a806f7ec4634d9bb71dea83c870526baef208fb441839d9fa):
```ts
export class Sample {
  private add: () => void;

  constructor(private remove: () => void) {
    this.add = () => {};
    this.remove = () => {};
  }
 // this as it stands is neither read or write, so needs work on reads / writes front or a special exception here
// but also must be considered in use readonly rule
  on(action: "add" | "remove"): void {
    this[action]();
  }
}
```


currently invalid in no unused private class members, but becoming valid since they are either read or write references:
```ts
class OnlyWrite {
	#usedOnlyInWrite = 5;

	method() {
			this.#usedOnlyInWrite = 212;
	}
}

class SelfUpdate {
	#usedOnlyToUpdateItself = 5;

	method() {
			this.#usedOnlyToUpdateItself++;
	}
}


class Foo {
	#usedOnlyInWrite = 5;
	method() {
			this.#usedOnlyInWrite = 42;
	}
}

class Foo {
	#usedOnlyInWriteStatement = 5;
	method() {
			this.#usedOnlyInWriteStatement += 42;
	}
}

class C {
	#usedOnlyInIncrement;

	foo() {
			this.#usedOnlyInIncrement++;
	}
}

```

**Comment by vladimir-ivanov** [comment]
•	Read = value flows out of the field.
	•	Write = value flows into the field.
	•	Compound/atomic updates (+=, ??=, ++) = Write-only.
	•	Simple assignment = Write-only.
	•	Explicit self-read (x = x + ...) = Read + Write.


Some examples:
	•	this.#x → Read
	•	let y = this.#x → Read
	•	console.log(this.#x) → Read
	•	this.#x + 1 → Read
	•	f(this.#x) → Read
	•	return this.#x → Read
	•	this.#x = 42 → Write
	•	class C { #x = 0 } → Write (initializer)
	•	this.#x = this.#x + 1 → Read + Write
	•	this.#x = this.#x ?? 3 → Read + Write
	•	this.#x ??= 3 → Write only (atomic update)
	•	this.#x += 1 → Write only
	•	this.#x |= 1 → Write only
	•	this.#x++ → Write only
	•	++this.#x → Write only
	•	this.#x = this.#x → Read + Write (useless)
	•	get #x() { return this.#y } → Read
	•	set #x(v) { this.#y = v } → Write
	•	const { x } = this → Read


### Based on the above definitions/ rules:
#### no-unused-private-class-members
	•	A field must have at least one meaningful read to be considered used.
	•	Write-only members (including +=, ??=, ++, etc.) are reported as unused.
#### prefer-readonly-class-properties
	•	A field that is written once (in constructor or initializer) and never written again but is read → should be marked readonly.
	•	Fields with writes beyond initialization → not readonly.
	•	Pure write-only fields (never read) → already caught by unused rule.

Think we need to introduce an additional flag on reads -> is_meaningful

**Comment by Conaclos** [comment]
One note regarding consistency: `noUnusedPrivateClassMembers` should also be consistent with `noUnusedVariables`/`noUnusedFunctionParameters`.

For example, `noUnusedVariables` considers `a ??= b` as a write-only operation.

**Comment by vladimir-ivanov** [comment]
@Conaclos  so please confirm/ or clarify if I am wrong:
 a ??= b to me is both read and write, but is_meaningful_read: false
whereas a = a ?? b is -> same both read and write but is_meaningful_read: false

it is the only one I need to sort out now to fully align the two without breaking any tests. 
Then once we have it in one centralised place we can easily make changes.

**Comment by vladimir-ivanov** [comment]
> One note regarding consistency: `noUnusedPrivateClassMembers` should also be consistent with `noUnusedVariables`/`noUnusedFunctionParameters`.
> 
> For example, `noUnusedVariables` considers `a ??= b` as a write-only operation.

was also thinking about that, would be nice if we can have some means to enforce it with code somehow (or shared tests at very least)

**Comment by Conaclos** [comment]
> a ??= b to me is both read and write, but is_meaningful_read: false
whereas a = a ?? b is -> same both read and write but is_meaningful_read: false

Yes.
However, to me some reads can be considered as "read for self-write", and then can be ignored.

Anyway, consistency is more important than considering or ignoring the read of `??=`. If it is the only read that we ignore, then I am fine with the idea of considering it as a read that makes a variable / member as used.

**Comment by vladimir-ivanov** [comment]
@Conaclos maybe going forward we can extract reusable pieces e..g is meaningful read to be shared in non class context

**Comment by vladimir-ivanov** [comment]
> > a ??= b to me is both read and write, but is_meaningful_read: false
> > whereas a = a ?? b is -> same both read and write but is_meaningful_read: false
> 
> Yes. However, to me some reads can be considered as "read for self-write", and then can be ignored.
> 
> Anyway, consistency is more important than considering or ignoring the read of `??=`. If it is the only read that we ignore, then I am fine with the idea of considering it as a read that makes a variable / member as used.

the concepts involved now are:

anything that have a `read` is marked as such, including both examples above.
also BOTH will end up being a `write` too.

Then I realised that these are not sufficient to make a call, we also need to separate meaningful (non useless) reads from any other reads.

hence the pr that I rated for is_meaningful_read on ClassMemberReference https://github.com/biomejs/biome/pull/7543
